docs/design/2020-06-01-global-kill.md

```diff
@@ -48,26 +54,67 @@ Bit 63 is always __ZERO__, making `connId` in range of non-negative int64, to be
 
-
 #### 4. serverId
-`serverId` is selected RANDOMLY by each TiDB instance on startup, and the uniqueness is guaranteed by PD(etcd). `serverId` should be larger or equal to 1, to ensure that high 32 bits of `connId` is always non-zero, and make it possible to detect truncation.
-
-On failure (e.g. fail connecting to PD, or all `serverId` are unavailable), we block any new connection.
+- `serverId` is selected RANDOMLY from `serverIds pool`_(see next)_ by each TiDB instance on startup, and the uniqueness is guaranteed by PD (etcd). `serverId` should be larger or equal to 1, to ensure that high 32 bits of `connId` is always non-zero, and make it possible to detect truncation.
 
-`serverId` is kept by PD with a lease (defaults to 12 hours, long enough to avoid brutally killing any long-run SQL). If TiDB is disconnected to PD longer than half of the lease (defaults to 6 hours), all connections are killed, and no new connection is accepted, to avoid running with stale/incorrect `serverId`. On connection to PD restored, a new `serverId` is acquired before accepting new connection.
+- `serverIds pool` is:
+  - All __UNUSED__ `serverIds` within [1, 2047] acquired from [`CLUSTER_INFO`](https://docs.pingcap.com/tidb/stable/information-schema-cluster-info#cluster_info) when 32 bits `connId` chosen.
+  - All `serverIds` within [2048, 2^22-1] when 64 bits `connId` chosen.
 
-On single TiDB instance without PD, a `serverId` of `1` is assigned.
+- On failure (e.g. fail connecting to PD, or all `serverId` are unavailable when 64 bits `connId` chosen), we block any new connection.
+- `serverId` is kept by PD with a lease (defaults to 12 hours, long enough to avoid brutally killing any long-run SQL). If TiDB is disconnected to PD longer than half of the lease (defaults to 6 hours), all connections are killed, and no new connection is accepted, to avoid running with stale/incorrect `serverId`. On connection to PD restored, a new `serverId` is acquired before accepting new connection.
+- On single TiDB instance without PD, a `serverId` of `1` is assigned.
 
 #### 5. local connId
-`local connId` is allocated by each TiDB instance on establishing connections incrementally.
+`local connId` is allocated by each TiDB instance on establishing connections:
 
-Integer overflow is ignored at this stage, as `local connId` should be long enough.
+- For 32 bits `connId`, `local connId` is possible to be integer-overflow and/or used up, especially on system being busy and/or with long running SQL. So we use a lock-free queue to maintain available `local connId`, dequeue on client connecting, and enqueue on disconnecting. When `local connId` exhausted, upgrade to 64 bits.
+
+- For 64 bits `connId`, allocate `local connId` by auto-increment, flip to zero if integer-overflow, and check `local connId` existed or not before allocation by maintaining a map of connection ids. At last, return _"Too many connections"_ error if exhausted.
```

Comment:
Why the implementations of 32 bits and 64 bits are different? Queue for 32 bits, but map for 64 bits?

Reply:
Maybe it is possible to grow the queue as needed. For example, when queue is drained (and we now that it allocated N elements), allocate id=N+1 to the new connection. When connection is closed, return this to the queue.

One thing to consider is that how queue can be compacted, as now it always occupies memory=max_peak_connections*N bytes, can be a source (i am not sure though) of memory leak: after peak workload memory is not released.

Reply:
> Why the implementations of 32 bits and 64 bits are different? Queue for 32 bits, but map for 64 bits?

`ConnID` allocation for 64 bits is just "auto-increment in the pool", as the pool size is big enough to avoid conflict (It needs more than 3 yrs to used up 2^40 integer in a 1w TPS instance). The map ([Server.clients](https://github.com/pingcap/tidb/blob/7e1533392030514440d27ba98001c374cdf8808f/server/server.go#L122)) in existed codes is just used for correctness with a trivial overhead. Improve doc to make here more clearly, PTAL @SunRunAway .

> 2^40 / (365 * 24 * 60 * 60 * 10000) = 3.48

Reply:
> Maybe it is possible to grow the queue as needed. For example, when queue is drained (and we now that it allocated N elements), allocate id=N+1 to the new connection. When connection is closed, return this to the queue.
> 
> One thing to consider is that how queue can be compacted, as now it always occupies memory=max_peak_connections*N bytes, can be a source (i am not sure though) of memory leak: after peak workload memory is not released.

En...I think it is not necessary to optimize memory usage ? Queue is for 32 bits `local connID`, and used at most 4MB memory (4B x 2^20) only.